### PR TITLE
 feat(moe): add DSv4 hash-based MoE routing (MOE-01-HASH)

### PR DESCRIPTION
**Pull request by leonardHONG**
## 📌 Description

Adds `flashinfer.fused_moe.hash_topk`, the DSv4 hash-based MoE expert routing kernel. DSv4-Pro hash-MoE layers select experts from a precomputed token-to-expert table `tid2eid` instead of running a dynamic top-k, so routing is an O(1) table lookup followed by a `sqrt(softplus(.))` score normalization. One warp processes one token.

Ported from SGLang's `deepseek_v4/hash_topk.cuh`. The Python signature matches `sglang.jit_kernel.deepseek_v4.hash_topk` so it works as a drop-in replacement.

### API

```python
from flashinfer.fused_moe import hash_topk

topk_weights, topk_ids = hash_topk(
    router_logits,                 # [num_tokens, num_routed_experts] float32
    input_ids,                     # [num_tokens] int64
    tid2eid,                       # [vocab, topk] int32
    num_fused_shared_experts=0,    # 0 or 1
    routed_scaling_factor=1.0,
    launch_with_pdl=True,
)
```

The routed weight for a selected expert is `sqrt(softplus(router_logits[token, expert])) / sum_over_routed`. When `num_fused_shared_experts == 1`, a shared-expert slot is appended with id `num_routed_experts` and weight `1 / routed_scaling_factor`.

### Implementation Notes

* SM-portable plain CUDA: warp-shuffle reduction plus a table gather, no tensor cores. PDL `griddepcontrol` is guarded on `__CUDA_ARCH__ >= 900`, with a non-PDL launch path for older architectures.
* `include/flashinfer/fused_moe/hash_topk.cuh` is framework-agnostic. PyTorch tensor handling and validation live in `csrc/fused_moe/hash_topk.cu`.
* Argument validation runs in both the Python API and the TVM-FFI layer, so direct FFI callers cannot bypass the contract: contiguity, same-device, dtype, `topk + num_fused_shared_experts <= 32`, `num_fused_shared_experts in {0, 1}`, and a positive finite `routed_scaling_factor` when a shared expert is fused.
* `num_tokens == 0` returns empty outputs without a zero-grid launch.
* A `hash_topk` trace template is added under `flashinfer/trace/templates/moe.py` and wired via `@flashinfer_api(trace=...)`.

 ## 🔍 Related Issues

  Implements the MOE-01-HASH cell of the DSv4 kernel tracker #3346.

### 🚀 Pull Request Checklist

Thank you for contributing to FlashInfer! Before we review your pull request, please make sure the following items are complete.

#### ✅ Pre-commit Checks

* [x] I have installed `pre-commit` by running `pip install pre-commit` or using my preferred method.
* [x] I have installed the hooks with `pre-commit install`.
* [x] I have run the hooks manually with `pre-commit run --all-files` and fixed any reported issues.

> If you are unsure about how to set up pre-commit, see the pre-commit documentation: https://pre-commit.com/.

#### 🧪 Tests

* [x] Tests have been added or updated as needed.
* [x] All tests are passing.

Validated on RTX 5090, SM120, CUDA 12.8:

* `tests/moe/test_hash_topk.py` — 44 cases covering `num_tokens in {1, 7, 64, 512, 4096}`, `num_routed_experts in {256, 384}`, shared-expert slot `{0, 1}`, the `topk + shared == 32` boundary, zero-token, PDL on/off, and argument validation against the SGLang torch reference.
* `tests/trace/test_fi_trace_template_consistency.py` — the `hash_topk` template passes the trace consistency checks.

## Reviewer Notes

The device kernel is a direct port of SGLang's `deepseek_v4/hash_topk.cuh`; correctness is checked against a torch reference of that kernel. This is a routing primitive, so there is no separate standalone benchmark — the `hash_topk` trace template provides the `fi_trace` benchmark definition.

## Release Notes

* **New Features**
  * Added `hash_topk` for DSv4 hash-based MoE routing, returning `topk_ids` and `topk_weights`.
  * Support for optional fused shared experts, including validation of routing scaling behavior.
  * Updated AOT/JIT generation and tracing so `hash_topk` is available and appears in MoE routing trace templates.

* **Tests**
  * Added CUDA-gated tests covering correctness vs. a reference, plus boundary cases and invalid-argument handling (including zero-token inputs).

### CODE REVIEW

File: flashinfer/trace/templates/moe.py [hunk 1]
```diff
@@ -2965,3 +2965,115 @@ def _b12x_fused_moe_reference(
...
+        "router_logits": Tensor(["num_tokens", "num_routed_experts"]),
+        "input_ids": Tensor(["num_tokens"], dtype="int64"),
+        "tid2eid": Tensor(["vocab", "topk"], dtype="int32"),
+        "num_fused_shared_experts": Scalar("int32"),
+        "routed_scaling_factor": Scalar("float32"),
+    },
+    outputs={
+        "topk_weights": Tensor(["num_tokens", "topk"], dtype="float32"),
+        "topk_ids": Tensor(["num_tokens", "topk"], dtype="int32"),
+    },
```
[the remark this answers is not in the document]
Comment (leonardHONG) [reply flashinfer/trace/templates/moe.py:3259]: The trace covers the routed path with num_fused_shared_experts=0, so [num_tokens, topk] matches the reference and  consistency checks pass; also Var takes no abbrev argument, so the suggested change would not import.

File: flashinfer/jit/hash_topk.py [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+
+def gen_hash_topk_module() -> JitSpec:
+    return gen_jit_spec(
+        "hash_topk",
+        [
+            jit_env.FLASHINFER_CSRC_DIR / "fused_moe" / "hash_topk.cu",
+            jit_env.FLASHINFER_CSRC_DIR / "fused_moe" / "hash_topk_jit_binding.cu",
+        ],
+        extra_cuda_cflags=["-lineinfo"],
+    )
```
[the remark this answers is not in the document]
Comment (leonardHONG) [reply flashinfer/jit/hash_topk.py:33]: added

File: flashinfer/fused_moe/hash_topk.py [hunk 3]
```diff
@@ -0,0 +1,223 @@
...
+    topk_fused = topk + num_fused_shared_experts
+
+    if num_routed_experts < 1:
+        raise ValueError(f"num_routed_experts must be >= 1, got {num_routed_experts}")
+    if topk < 1:
+        raise ValueError(f"topk (tid2eid.shape[1]) must be >= 1, got {topk}")
+    if input_ids.shape[0] != num_tokens:
+        raise ValueError(
+            f"input_ids length ({input_ids.shape[0]}) must equal num_tokens ({num_tokens})"
+        )
```
[the remark this answers is not in the document]
Comment (leonardHONG) [reply flashinfer/fused_moe/hash_topk.py:77]: Caller's contract, matching the SGLang source — a host-side range check would break CUDA-graph capture.

**Review by aleozlx — APPROVED** [review]
lgtm
